### PR TITLE
fix(evmrpc): fail fast on pruned heights for debug_trace* (PLT-975)

### PR DESCRIPTION
**Pull request by amir-deris**
## Describe your changes and provide context

`debug_trace*` endpoints could silently return empty results or panic when block, receipt, or state data had been pruned. Guards also ran after the trace semaphore was acquired, so pruned requests could block on concurrency limits instead of failing immediately.

This PR adds retention checks before trace work begins and returns explicit errors when data is unavailable:

- **`EnsureTraceHeightAvailable`** — for replay endpoints (`debug_traceTransaction`, `debug_traceBlockByNumber`, etc.): verifies block, parent block (height−1 for validator/state replay), receipt, and state retention. Fetches watermarks once per guard (single `Status` snapshot).
- **`EnsureTraceCallHeightAvailable`** — for `debug_traceCall`: verifies block and state only (no receipt check, since TraceCall never reads receipts). Same single-snapshot watermark fetch.
- **`ErrReceiptPruned`** — new sentinel wrapping `ErrNotFound` so `eth_getTransactionReceipt` still returns `null` for pruned receipts while trace guards can distinguish pruned from missing.
- **`latestTraceHeight`** — resolves `latest`/`pending`/`safe`/`finalized` tags via the watermark's safe latest instead of the raw app tip, avoiding intermittent errors when receipts/state lag the tip by a block.
- **Guard ordering** — retention checks intentionally run before the trace semaphore so pruned requests fail fast without consuming a slot; hash-based endpoints may perform Tendermint block lookups outside the concurrency limiter as part of those guards (reverses the previous semaphore-first invariant).
- **`Backend.BlockByNumber`** — uses shared `getBlockNumber` + watermark resolution instead of a separate `ConvertBlockNumber` path; `pending` now resolves like `latest` per EVM RPC spec instead of panicking.
- **Minor fixes** — empty tx list no longer reports a block trace cache hit (avoids serving a stale empty result when tx enumeration failed); nil EVM tx in block filtering is skipped safely.

**Known limitation:** tx-by-hash trace guards need a resolvable receipt (or index); once a receipt is fully TTL-reclaimed, `ErrNotFound` is indistinguishable from an unknown hash and the request proceeds into the trace path before failing downstream.

Fixes PLT-975.

## Testing performed to validate your change

- [x] Added unit tests for watermark trace guards (`EnsureTraceHeightAvailable`, `EnsureTraceCallHeightAvailable`, parent block floor, genesis initial height, SS-disabled edge cases)
- [x] Added integration-style tests for trace endpoints rejecting pruned heights before semaphore acquisition
- [x] Added tests for `latest` tag guard matching block resolution via safe latest watermark
- [x] Added tests for receipt lookup error propagation (`ErrReceiptPruned`, store errors, `ErrNotFound` fallthrough)
- [x] Updated receipt store tests to expect `ErrReceiptPruned` below retention floor
- [x] Updated trace semaphore ordering tests (`TraceBlockByNumber`, `TraceBlockByHash`, `TraceCall`)

### CODE REVIEW

File: evmrpc/simulate.go [hunk 1]
```diff
@@ -410,26 +410,12 @@ func (b *Backend) ChainDb() ethdb.Database {
...
-	case rpc.PendingBlockNumber.Int64():
-		panic("tracing on pending block is not supported")
-	}
-	return blockNum
-}
-
 func (b Backend) BlockByNumber(ctx context.Context, bn rpc.BlockNumber) (*ethtypes.Block, []tracersutils.TraceBlockMetadata, error) {
-	blockNum := b.ConvertBlockNumber(bn)
-	tmBlock, err := blockByNumberRespectingWatermarks(ctx, b.tmClient, b.watermarks, &blockNum, 1)
+	blockNumberPtr, err := getBlockNumber(ctx, b.tmClient, bn)
```
[the remark this answers is not in the document]
Comment (amir-deris) [reply evmrpc/simulate.go:414]: Yes, debug_traceBlockByNumber("pending") used to panic; it now traces the watermark safe latest. Sei's own RPC spec already says pending should behave like latest in `sei-chain/evmrpc/AGENTS.md`. The trace guard path already treated pending like latest via resolveDebugTraceBlockNumber → latestTraceHeight. The old ConvertBlockNumber panic meant guard and execution disagreed: guard could accept "pending", then BlockByNumber would panic. This change fixes that mismatch, it doesn't introduce a new semantic.

File: evmrpc/tracers.go [hunk 2]
```diff
@@ -101,50 +102,119 @@ func (api *DebugAPI) prepareTraceContext(ctx context.Context) (context.Context,
...
-	if err != nil || receipt == nil {
-		return nil
+	return api.guardHistoricalDebugTraceHeight(ctx, endpoint, height)
+}
+
+func (api *DebugAPI) guardTraceRequestByTxHash(ctx context.Context, endpoint string, hash common.Hash) error {
+	if api.keeper != nil {
+		rcpt, err := api.keeper.GetReceipt(api.ctxProvider(LatestCtxHeight), hash)
+		if err != nil {
+			if errors.Is(err, receipt.ErrReceiptPruned) || !errors.Is(err, receipt.ErrNotFound) {
```
[the remark this answers is not in the document]
Comment (amir-deris) [reply evmrpc/tracers.go:146]: Agree — this is a real coverage gap, not a regression.

What the PR closes: fail-fast when a receipt is still retrievable:

litt lazy-TTL window → ErrReceiptPruned (handled at L121)
default store, receipt bytes still present → rcpt.BlockNumber → EnsureTraceHeightAvailable
What it can't close: once the receipt is physically gone, GetReceipt returns plain ErrNotFound with no resolvable height. Without another source (tx index, block enumeration, etc.) we can't distinguish "unknown hash" from "receipt TTL-reclaimed," so the guard correctly falls through to lookback at latestTraceHeight. That's intentional — see TestGuardTraceRequestByTxHashReceiptLookupErrors/ErrNotFound falls through to lookback and the Known limitation in the PR description.

On the heuristic suggestion: returning "receipt not found; receipts below height %d have been pruned" when EarliestVersion() > blockEarliest would fire for genuinely unknown recent hashes too, since we still don't know which block the tx lived at. Misleading for valid ErrNotFound cases, so I'd skip it here.

Net: this path is strictly better than before (old code returned nil on any receipt error, with no watermark checks). Fully TTL-reclaimed receipts still proceed into trace and fail downstream — worth a follow-up issue if we want to avoid consuming a semaphore slot, but out of scope for PLT-975.

[the remark this answers is not in the document]
Comment (amir-deris) [reply evmrpc/tracers.go:146]: Valid observation on the behavior change, but ErrNotConfigured isn't reachable on a normally configured node — app.go always constructs a receipt store at startup and panics on failure, so production keepers always have one.

ErrNotConfigured means keeper.receiptStore == nil, which is a misconfiguration/test-only state. Propagating it here is intentional and consistent with eth_getTransactionReceipt, which also returns an RPC error (not null) on this sentinel. The old guard swallowed all receipt errors including ErrNotConfigured, which masked misconfiguration and let trace proceed into a path that needs receipts anyway.

**Comment by amir-deris** [comment]
@seidroid review

**Comment by amir-deris** [comment]
cancelling this PR as the issue with debug trace was resolved by another PR.
